src/test/parse-fail/qquote-2.rs

```diff
@@ -10,54 +10,47 @@
 
 // compile-flags: -Z parse-only
 
-// ignore-test Can't use syntax crate here
+// ignore-cross-compile
 
-#![feature(quote)]
+#![feature(quote, rustc_private)]
 
 extern crate syntax;
 
-use syntax::diagnostic;
 use syntax::ast;
 use syntax::codemap;
-use syntax::parse::parser;
-use syntax::print::*;
-
-trait fake_ext_ctxt {
-    fn cfg() -> ast::CrateConfig;
-    fn parse_sess() -> parse::parse_sess;
-    fn call_site() -> span;
-    fn ident_of(st: &str) -> ast::ident;
+use syntax::parse;
+use syntax::print::pprust;
+
+trait FakeExtCtxt {
+    fn call_site(&self) -> codemap::Span;
+    fn cfg(&self) -> ast::CrateConfig;
+    fn ident_of(&self, st: &str) -> ast::Ident;
+    fn name_of(&self, st: &str) -> ast::Name;
+    fn parse_sess(&self) -> &parse::ParseSess;
 }
 
-type fake_session = parse::parse_sess;
-
-impl fake_ext_ctxt for fake_session {
-    fn cfg() -> ast::CrateConfig { Vec::new() }
-    fn parse_sess() -> parse::parse_sess { self }
-    fn call_site() -> span {
-        codemap::span {
+impl FakeExtCtxt for parse::ParseSess {
+    fn call_site(&self) -> codemap::Span {
+        codemap::Span {
             lo: codemap::BytePos(0),
             hi: codemap::BytePos(0),
-            expn_id: codemap::NO_EXPANSION
+            expn_id: codemap::NO_EXPANSION,
         }
     }
-    fn ident_of(st: &str) -> ast::ident {
-        self.interner.intern(st)
+    fn cfg(&self) -> ast::CrateConfig { Vec::new() }
+    fn ident_of(&self, st: &str) -> ast::Ident {
+        parse::token::str_to_ident(st)
     }
+    fn name_of(&self, st: &str) -> ast::Name {
+        parse::token::intern(st)
+    }
+    fn parse_sess(&self) -> &parse::ParseSess { self }
 }
 
-fn mk_ctxt() -> fake_ext_ctxt {
-    parse::new_parse_sess(None) as fake_ext_ctxt
-}
-
-
 fn main() {
-    let cx = mk_ctxt();
-
-    let stmt = quote_stmt!(cx, let x isize = 20;); //~ ERROR expected end-of-string
-    check_pp(*stmt,  pprust::print_stmt, "");
-}
+    let cx = parse::new_parse_sess();
 
-fn check_pp<T>(expr: T, f: |pprust::ps, T|, expect: str) {
-    panic!();
+    quote_expr!(&cx, let x isize = 20;).and_then(|expr| { //~ ERROR expected end-of-string
```

Comment:
@alexcrichton do you have any thoughts on how this can be made to fail in the expected way? it currently compiles without issue and then crashes at runtime